timeboost-types/src/bundle.rs

```diff
@@ -0,0 +1,463 @@
+use committable::{Commitment, Committable, RawCommitmentBuilder};
+use serde::{Deserialize, Serialize};
+use timeboost_crypto::KeysetId;
+
+use crate::{Bytes, Epoch, SeqNo};
+
+const DOMAIN: &str = "TIMEBOOST_BID";
+
+#[derive(Debug, Clone)]
+pub enum BundleVariant {
+    Regular(Bundle),
+    Priority(PriorityBundle<Signed>),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Bundle {
+    chain: ChainId,
+    epoch: Epoch,
+    data: Bytes,
+    kid: Option<KeysetId>,
+    hash: [u8; 32],
+}
+
+impl Bundle {
+    pub fn new(chain: ChainId, epoch: Epoch, data: Bytes, kid: Option<KeysetId>) -> Self {
+        let mut this = Self {
+            chain,
+            epoch,
+            data,
+            kid,
+            hash: [0; 32],
+        };
+        this.update_hash();
+        this
+    }
+
+    fn update_hash(&mut self) {
+        let digest = self.commit();
+        self.hash = digest.into();
+    }
+}
+
+impl Bundle {
+    pub fn chain_id(&self) -> ChainId {
+        self.chain
+    }
+
+    pub fn epoch(&self) -> Epoch {
+        self.epoch
+    }
+
+    pub fn data(&self) -> &Bytes {
+        &self.data
+    }
+
+    pub fn kid(&self) -> Option<KeysetId> {
+        self.kid
+    }
+
+    pub fn digest(&self) -> &[u8; 32] {
+        &self.hash
+    }
+
+    pub fn set_data(&mut self, d: Bytes) {
+        self.data = d;
+        self.update_hash()
+    }
+
+    #[cfg(feature = "arbitrary")]
+    pub fn arbitrary(u: &mut arbitrary::Unstructured<'_>) -> arbitrary::Result<Bundle> {
+        use alloy_rlp::Encodable;
+        use arbitrary::Arbitrary;
+        let t: Transaction = loop {
+            let candidate: Transaction = arbitrary::Arbitrary::arbitrary(u)?;
+            if let TxEnvelope::Eip4844(ref eip4844) = candidate.0 {
+                if eip4844.tx().clone().try_into_4844_with_sidecar().is_ok() {
+                    // Avoid generating 4844 Tx with blobs of size 131 KB
+                    continue;
+                }
+            }
+            break candidate;
+        };
+
+        let mut d = Vec::new();
+        t.encode(&mut d);
+        let c = ChainId::default();
+        let e = Epoch::now() + bool::arbitrary(u)? as u64;
+        let encoded = ssz::ssz_encode(&vec![&d]);
+        let k = None;
+        let b = Bundle::new(c, e, encoded.into(), k);
+
+        Ok(b)
+    }
+}
+
+impl Committable for Bundle {
+    fn commit(&self) -> Commitment<Self> {
+        RawCommitmentBuilder::new("PriorityBundle")
+            .field("chain", self.chain_id().commit())
+            .field("epoch", self.epoch().commit())
+            .var_size_field("data", self.data())
+            .field("keysetid", self.kid().unwrap_or_default().commit())
+            .finalize()
+    }
+}
+
+pub trait State {}
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Signed {
+    signature: Signature,
+}
+pub struct Unsigned {}
```

Comment:
```suggestion
pub struct Unsigned(());
```

Maybe the private field can be added to ensure that `Unsigned` values can only be constructed in here?